### ISSUE TITLE
Support custom "reverse requests" from debug adapters

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#165165]
Debug adapters can send events that are not part of the DAP spec that can be handled by extensions by subscribing to [`vs.debug.onDidReceiveDebugSessionCustomEvent`](https://code.visualstudio.com/api/references/vscode-api#debug.onDidReceiveDebugSessionCustomEvent).

But events are not the only thing that DAs can send, they can also send ["reverse requests"](https://microsoft.github.io/debug-adapter-protocol/specification#reverse-requests) such as `runInTerminalRequest`.

It's not currently possible to create custom reverse requests because there is no hook like `onDidReceiveDebugSessionCustomEvent` for handling custom requests. This can be useful when a debug adapter needs to get some information from the extension - for example if a debug adapter is launching a web app and needs to expose a URI through `vscode.env.asExternalUri()` and get back the public URI. Right now this has to be done by firing an event from the DA and then having the extension send back another event (which feels awkward given reverse requests already exist in DAP).

@weinand I filed this here because I think it's a VS Code API that's missing and not something for the DAP spec, but if it makes more sense somewhere else like (https://github.com/Microsoft/debug-adapter-protocol), let me know and I can re-raise it. Thanks!

**Comment by roblourens** [comment]
I'm not sure how it would work to have a DA send a reverse request that is not handled by vscode but by some extension. I guess you're asking for an API that lets an extension install a handler for a certain request type? Does that mean we need a mechanism for custom capabilities too?

**Comment by DanTup** [comment]
> I'm not sure how it would work to have a DA send a reverse request that is not handled by vscode but by some extension. I guess you're asking for an API that lets an extension install a handler for a certain request type?

Yep, something like that sounds reasonable (I thought a single handler would work, but in hindsight that won't be easy to fall through to other extensions for requests you don't know, so providing the specific request type makes sense).


> Does that mean we need a mechanism for custom capabilities too?

That would probably be useful too, although I'm already stitching a bunch of flags into the launch config so that's less of a concern to me.

(ofc, if asExternalUri is something other DAs might be interested in, making that a first-class reverse-request in DAP would also suit me, though it's not as extensible)

**Comment by connor4312** [comment]
I agree that this would be useful.
